### DISCUSSION TITLE
Best practices for creating RepresentationFormat from IDL

### DISCUSSION THREAD

**Discussion by Idan-Neeman** [OpenDDS/OpenDDS#4232] [Q&A]
I am looking for advice on the recommended workflow for generating an OpenDDS::DCPS::RepresentationFormat from an IDL file, which can properly serialize and deserialize the defined data types or bytes.

The steps I have tried so far:
1. Run tao_idl on the IDL file to generate C++ type code.
2. Run opendds_idl on the same IDL to generate OpenDDS type support code.
3. In my application, I attempt to construct the RepresentationFormat using the TypeSupportImpl code.

The main issue I'm running into is that the TypeSupport generated code doesn't seem to contain the full set of needed methods like make_format(), decode_from_bytes().

Any advice, or example code for properly creating the RepresentationFormat would be greatly appreciated.

**Comment by iguessthislldo** [comment]
As I mentioned [very recently](https://github.com/OpenDDS/OpenDDS/pull/4122#issuecomment-1678188763) (...coincidence?), we don't support using those methods on the CDR representation kinds yet. They were added not that long ago specifically for another project to use JSON sample conversion code in Java as well as in C++. The possibility of using them with CDR was an added bonus and it wouldn't be hard to finish support for that, but we just haven't gotten around to it yet. Otherwise you'd have to use the internal `Serializer` like I mentioned in the link.

**Reply by Idan-Neeman** [reply]
I appreciate your help 👍 
Given the current situation, does this imply that I'm unable to deserialize the CDR bytes data (of sample) into a DynamicData object?

**Reply by iguessthislldo** [reply]
You can if you use the internal [`DynamicDataXcdrReadImpl`](https://github.com/OpenDDS/OpenDDS/blob/master/dds/DCPS/XTypes/DynamicDataXcdrReadImpl.h) class along with the `Serializer`. Using it would have a similar setup as I mentioned [here](https://stackoverflow.com/a/72016334/9189546) where I serialize from and deserialize to a C++ type, except you'd pass the `Serializer` to `DynamicDataXcdrReadImpl` instead of doing `serializer >> idl_value`. Then you can use `DynamicDataXcdrReadImpl` to get values from the CDR stream. I can give a more complete example if you need one.

**Reply by Idan-Neeman** [reply]
Hey, I would appreciate if you could provide a complete example demonstrating how to use the internal DynamicDataXcdrReadImpl class to deserialize sample from a CDR bytes. Thank you 🥇

**Reply by iguessthislldo** [reply]
Alright. If you haven't already figured it out, here's an example:

```c++
#include <dds/DCPS/Serializer.h>
#include <dds/DCPS/XTypes/DynamicDataXcdrReadImpl.h>

\\ ...

// What's needed
DDS::DynamicType_var dt = /* type to use */;
const char* const data = /* CDR Data */;
const size_t data_size = /* size of CDR size */;

// Get the DynamicData
ACE_Message_Block mb(data_size);
mb.copy(data, data_size);
OpenDDS::DCPS::Serializer ser(&mb, OpenDDS::DCPS::Encoding::KIND_XCDR2);
DDS::DynamicData_var dd = new OpenDDS::XTypes::DynamicDataXcdrReadImpl(ser, dt);
```

This is making a few assumptions:
- [You can get the `DynamicType`](https://opendds.readthedocs.io/en/latest-release/devguide/xtypes.html#obtaining-dynamictype-and-registering-typesupport)
- The CDR is in the same endianess and doesn't have an encapsulation header.

**Reply by iguessthislldo** [reply]
One more thing I should mention: a downside of using `DynamicDataXcdrReadImpl` is that not only is it an internal class, but we're actively looking at getting rid of it because right now writing and reading CDR doesn't happen with the same `DynamicData` implementation. If you're willing, you could take this and implement `decode_from_bytes` on [`DynamicTypeSupport`](https://github.com/OpenDDS/OpenDDS/blob/master/dds/DCPS/XTypes/DynamicTypeSupport.h#L83). If you submitted that as a pull request we'd have to keep it updated as we make changes internally.

**Reply by Idan-Neeman** [reply]
I appreciate your assistance with this matter. I've been working on incorporating the example code, and I've come across a challenge when attempting to execute the function:

```cpp
TestODS::ShapeTypeTypeSupport_var type_support = new TestODS::ShapeTypeTypeSupportImpl;
DDS::DynamicType_var dt = type_support->get_type();
```

Unfortunately, this line yields a NULL pointer. After delving deeper into the issue, I discovered that the function type_identifier_to_dynamic returns 0 due to the ti.kind() being evaluated as TK_NONE.

Any insights or guidance you could provide to help resolve this would be greatly appreciated.

**Reply by iguessthislldo** [reply]
It can do that if [complete type information hasn't been enabled](https://opendds.readthedocs.io/en/master/devguide/xtypes.html#enabling-use-of-completetypeobjects).

**Reply by Idan-Neeman** [reply]
Yeah, that was the problem :)
How can I access the representation data (fields name -> value) or structure of the data using the 'dd' variable? 
My objective is to output the sample in the form of a JSON object or a map.

Thank you!

**Reply by iguessthislldo** [reply]
[The same page of the Developer's Guide](https://opendds.readthedocs.io/en/master/devguide/xtypes.html#reading-basic-types) covers how to use a `DynamicData`. However if JSON is the only thing you need, then it might be easier to use our builtin JSON support in `encode_to_string`.

From CDR data to JSON would look like:

```c++
TestODS::ShapeTypeTypeSupport_var type_support = new TestODS::ShapeTypeTypeSupportImpl;
const char* const data = /* CDR Data */;
const size_t data_size = /* size of CDR size */;

// Deserialize
ACE_Message_Block mb(data_size);
mb.copy(data, data_size);
OpenDDS::DCPS::Serializer ser(&mb, OpenDDS::DCPS::Encoding::KIND_XCDR2);
TestODS::ShapeType value;
if (!(ser >> value)) {
  // Deserialization failed
}

// Make JSON
OpenDDS::DCPS::RepresentationFormat_var format =
  type_support->make_format(OpenDDS::DCPS::JSON_DATA_REPRESENTATION);
CORBA::String_var json;
if (type_support->encode_to_string(value, json, format) != DDS::RETCODE_OK) {
  // JSONify failed
}
printf("%s\n", json.in());
```

I haven't tested this like the other example, so something might be wrong here.

Alternatively if you just need a human friendly form, but not necessarily JSON, there's also `print_dynamic_data`. Unfortunately there's not an easy way to get a `DynamicData` into JSON yet.

**Comment by mitza-oci** [comment]
There is one in OpenDDS already: See
OpenDDS::DCPS::DynamicSample::deserialize().
I suggest stepping through this in a debugger to see how it works.

**Reply by iguessthislldo** [reply]
Yes, but I didn't point to that one because it's making a new `ACE_Message_Block` in a way that's particular to `DynamicSample` and even though it's not hard to make, it's assuming the `Serializer` already exists.
